File: truth/backend/src/truth_service/models/query.py
"""
The query payload behind the register's table - free text, structured restrictions, ordering and the window.

:date: 2026-09-08
"""
# ----- IMPORTS ----- #

from pydantic import BaseModel, ConfigDict, Field

from skyscanner_models.query import FilterCondition, SortSpecification

from truth_service.models.assumption import AssumptionRowResponse

# ----- CONSTS ----- #

DEFAULT_LIMIT: int = 50
MAX_LIMIT: int = 1000

# ----- CLASSES ----- #


class AssumptionQuery(BaseModel):
    """
    Everything the table is asking the register for at once.

    The window is stated as an offset and a limit rather than as a page, so that the same payload serves the
    table, an export walking the whole answer and anything else reading in blocks.
    """

    model_config = ConfigDict(populate_by_name=True)

    search: str | None = Field(default=None, description="Free text matched against the indexed assumption text")
    industry: str | None = Field(default=None, description="Industry, by identifier or name, to narrow to")
    schema_key: str | None = Field(
        default=None,
        alias="schema",
        description="Declaration, by identifier or name, to narrow to",
    )
    filters: list[FilterCondition] = Field(default_factory=list, description="Structured attribute restrictions")
    sort: list[SortSpecification] = Field(default_factory=list, description="Ordering applied to the answer")
    offset: int = Field(default=0, ge=0, description="Amount of assumptions skipped before collecting")
    limit: int = Field(default=DEFAULT_LIMIT, ge=1, le=MAX_LIMIT, description="Largest amount to return")
    include_total: bool = Field(default=True, description="Whether the answer counts the whole match as well")


class AssumptionPage(BaseModel):
    """
    One window of an answer, together with how many assumptions the whole of it holds.
    """

    model_config = ConfigDict(populate_by_name=True)

    rows: list[AssumptionRowResponse] = Field(default_factory=list, description="The assumptions of this window")
    total: int = Field(default=0, description="How many assumptions matched, before the window was taken")
    offset: int = Field(default=0, description="Where this window starts")
    limit: int = Field(default=DEFAULT_LIMIT, description="How large this window was asked to be")


class FacetResponse(BaseModel):
    """
    Every value one attribute is known to hold, which is the vocabulary its filter offers to pick from.
    """

    model_config = ConfigDict(populate_by_name=True)

    key: str = Field(description="Attribute the values belong to")
    values: list[str] = Field(default_factory=list, description="The values that attribute is known to hold")
    truncated: bool = Field(default=False, description="Whether the register holds more values than were read")

File: truth/frontend/src/models/query.ts
/**
 * The question the table asks the register, and the answer it gets back.
 *
 * Every part of this is answered by the service rather than in the browser. The register is expected to hold
 * far more assumptions than a browser can be handed, so what arrives is one window of an answer and the size
 * of the whole of it - never the register itself.
 */

import type { FilterCondition, SortSpecification } from '@truth-platform/core-ui'

import type { AssumptionRow } from './assumption'

/** Everything the table is asking at once. */
interface AssumptionQuery {
  search: string | null
  /** The industry the answer is narrowed to, by identifier or by name, or nothing for the whole register. */
  industry: string | null
  filters: FilterCondition[]
  sort: SortSpecification[]
  offset: number
  limit: number
  /**
   * Whether the answer counts the whole match as well as returning the window.
   *
   * A table needs the count to draw its pager. Anything walking the answer in blocks - an export, say - knows
   * it has reached the end when a block comes back short, and counting again for every block doubles the work
   * of the whole walk for something already known.
   */
  include_total?: boolean
}

/** One window of an answer, and how many assumptions the whole of it holds. */
interface AssumptionPage {
  rows: AssumptionRow[]
  total: number
  offset: number
  limit: number
}

/** Every value one column is known to hold, which is what its filter offers to pick from. */
interface Facet {
  key: string
  values: string[]
  /** Whether the register holds more values than were gathered, so the filter says what it is offering. */
  truncated: boolean
}

export type { AssumptionPage, AssumptionQuery, Facet }
